File: lisp_parser.c
#include "lisp_parser.h"

#define SEGS_MAX 50

LispCmd* lisp_parser(char* lisp_str) {
  int depth = 0;
  int instring= 0;
  int lastpos =0;

  int segs_number=0;

  char**segs =(char**)malloc(SEGS_MAX*sizeof(char*));
  
  LispCmd *lisp_cmd=NULL;
  int i=0;

  for(i=0;i<strlen(lisp_str);i++ ) {
    if (lisp_str[i] != ' ' && lisp_str[i] != '(' ) {
      if(depth == 0) {
        printf("syntax error %d\n",i);
        return NULL;
      }
    }

    if (lisp_str[i]) == '('){
      depth+=1;
      lastpos=i;
    }
    
    if (lisp_str[i] == ')') {
      if(lastpos < i) {
        char*tmp = (char*)malloc((i-lastpos+1+1)*sizeof(char));
        strncpy(tmp, lisp_str +lastpos+1, i-lastpos+1 );
        tmp[ i-lastpos+1 ] = '\0';
        if(segs_number < SEGS_MAX) { 
          segs[segs_number] = tmp;
          segs_number+=1;
        }else{
          break;
        }
      }

      depth-=1;
    }
    
    if (depth > 0 ) {
      if (lisp_str[i] == ' ') {
        if(instring == 0) {
          char*tmp = (char*)malloc((i-lastpos+1+1)*sizeof(char));
          strncpy(tmp, lisp_str +lastpos+1, i-lastpos+1 );
          tmp[ i-lastpos+1 ] = '\0';
          if(segs_number < SEGS_MAX) { 
            segs[segs_number] = tmp;
            segs_number+=1;
          }else{
            break;
          }
        }
      }
    }
    
    if (lisp_str[i] == '"') {
      if(instring == 0) {
        instring+=1;
      }else if(instring > 0 ){
        instring-=1;
      }
    }


  }

  if(depth > 0) {
    for(i=0;i<segs_number;i++){
      free(segs[i]);
    }
    free(segs);
    printf("syntax error ,unexcepted closure\n");
    return NULL;
  }
  
  if( instring > 0) {
    for(i=0;i<segs_number;i++){
      free(segs[i]);
    }
    free(segs);
    printf("syntax error , string quato errors\n");
    return NULL;
  }
  

  if(segs_number < 1) {
    for(i=0;i<segs_number;i++){
      free(segs[i]);
    }
    free(segs);
    printf("unknown error\n");
    return NULL;
  }
  
  lisp_cmd = (LispCmd*)malloc(sizeof(LispCmd));
  lisp_cmd->Func = segs[0];
  lisp_cmd->Args = NULL;
  lisp_cmd->Argc = 0;

  if(segs_number -1 > 0) {
    lisp_cmd->Args = (CmdArgs*)malloc( (segs_number-1)*sizeof(CmdArg));
    lisp_cmd->Argc = segs_number-1;
  }
  
  for(i=1;i<segs_number;i++) {

    if(segs[i][0] == '"') {
      lisp_cmd->Args[i].Type = 'S';
      strncpy(lisp_cmd->Args[i].Value, segs[i]+1, strlen(segs[i])-1-1 );

    }else if(strcmp(segs[i],"true") == 0 || strcmp(segs[i],"false") == 0 ) {
      lisp_cmd->Args[i].Type = "B";
      strcpy(lisp_cmd->Args[i],segs[i]);

    }else if(strchr(segs[i],'.') != NULL){
      lisp_cmd->Args[i].Type= "F";
      strcpy(lisp_cmd->Args[i],segs[i]);
    }else {
      lisp_cmd->Args[i].Type= "I";
      strcpy(lisp_cmd->Args[i],segs[i]);      
    }
    
  }


  for(i=0;i<segs_number;i++){
      free(segs[i]);
  }  
  free(segs);
  
  return lisp_cmd;
}

//------------------------------------------------------------------------------
char *CmdArg_GetStr(CmdArg*arg) {
  if(arg->Type != 'S') {
    return NULL;
  }

  return arg->Value;
}

double CmdArg_GetFloat(CmdArg*arg) {
  double ret;
  ret = 0.0;
  if(arg->Type != 'F') {
    return ret;
  }
  
  ret = strtod(arg->Value,NULL);
  
  return ret;

}


int CmdArg_GetInt(CmdArg*arg) {
  int ret;
  ret = 0;
  if(arg->Type != 'I') {
    return ret;
  }

  ret = atoi(arg->Value);
  return ret;
}

bool CmdArg_GetBoolean(CmdArg*arg) {
  if(arg->Type != "B") {
    return false;
  }

  if( strcmp( arg->Value,"true") == 0)  {
    return true;
  }

  if( strcmp( arg->Value,"false") == 0)  {
    return false;
  }
  
  return false;

}

File: pico8/lisp_parser.lua
local lisp = {}

local function starts_with(str, start)
   return str:sub(1, #start) == start
end

local function ends_with(str, ending)
   return ending == "" or str:sub(-#ending) == ending
end

function lisp.parser(lisp_str)
	
	local segs = {}
	local depth = 0
	local instring = 0
	local lastpos = 0
	local segs_number=0
	local LispCmd = {Func="",Args={}}

	for i = 1, #lisp_str do
    local c = lisp_str:sub(i,i)
		if c ~= ' ' and c ~= '(' then
			if depth ==0 then
				print("syntax error ",i,lisp_str)
				return
			end
		end

		if c == '(' then
			depth = depth +1
			lastpos = i
		end
		
		if c == ')' then
			if lastpos < i then	
				local tmp = lisp_str:sub(lastpos+1,i-1)
				segs[segs_number] = tmp
				segs_number = segs_number+1
			end

			depth = depth -1
		end
		
		if depth > 0 then
			if c == ' ' then
				if instring == 0 then
					local tmp = lisp_str:sub(lastpos+1,i-1)
					
					segs[segs_number] = tmp
					segs_number = segs_number + 1
					
					lastpos = i
				end
			end
		end
		
		if c == '"' then
			if instring == 0 then
				instring = instring +1
			elseif instring > 0 then
				instring = instring -1
			end
		end	
	end
	
	if depth > 0 then
		print("syntax error ,unexcepted closure")
		return nil
	end

	if instring > 0 then
		print("syntax error , string quato errors")
		return nil
	end

	if segs_number < 1 then
		print("unknown error")
		return nil
	end
	
	LispCmd.Func = segs[0]

	for i,v in ipairs(segs) do
		if starts_with(v,'"') and ends_with(v,'"') then
			segs[i] = segs[i]:sub(2,#segs[i]-1)
		end
	end
	LispCmd.Args = segs
	
	return LispCmd
end

return lisp

--[[

function foo(a,b)
	print("i am foo", a,  b)
end

funcs = {}
funcs["foo"] = foo

l = lisp_parser("(foo 1 \"hah\")")

funcs[l.Func](l.Args[1],l.Args[2])

]]

